master/internal/rm/kubernetesrm/job.go

```diff
@@ -143,284 +140,363 @@ func newPod(
+			if err := j.startPodLogStreamers(); err != nil {
 				return err
 			}
 		}
 	} else {
-		if err := p.createPodSpecAndSubmit(); err != nil {
+		if err := j.createSpecAndSubmit(); err != nil {
 			return fmt.Errorf("creating pod spec: %w", err)
 		}
 	}
 	return nil
 }
 
-func (p *pod) finalize() {
-	p.kill()
-	p.finalizeTaskState()
+func (j *job) finalize() {
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	// If an error occurred during the lifecycle of the pods, we need to update the scheduler
+	// and the task handler with new state.
+	if j.container.State != cproto.Terminated {
+		j.kill()
+		j.syslog.Warnf("killed job after our handler exited unexpectedly")
+		j.container = j.container.Transition(cproto.Terminated)
+		j.informTaskResourcesStopped(sproto.ResourcesError(
+			sproto.TaskError,
+			errors.New("pod handler exited while pod was running"),
+		))
+	}
 }
 
-func (p *pod) podStatusUpdate(updatedPod *k8sV1.Pod) (cproto.State, error) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+func (j *job) jobStatusUpdate(updatedJob *batchv1.Job) (cproto.State, error) {
+	j.mu.Lock()
+	defer j.mu.Unlock()
 
-	if p.container.State == cproto.Terminated {
-		return p.container.State, nil
+	conds := updatedJob.Status.Conditions
+	if len(conds) == 0 {
+		return j.container.State, nil
 	}
 
-	p.pod = updatedPod
+	for _, cond := range conds {
+		if cond.Status != k8sV1.ConditionTrue {
+			continue
+		}
+
+		switch cond.Type {
+		case batchv1.JobComplete:
+			j.container.State = cproto.Terminated
+			j.informTaskResourcesStopped(sproto.ResourcesStopped{Failure: nil})
+			return cproto.Terminated, nil
+
+		case batchv1.JobFailed:
+			if j.firstPodExit == nil {
+				j.firstPodExit = &podExitInfo{
+					code: 1025,
+					msg:  "job exited with a failure but we don't have pod-level detail",
+				}
+			}
+			j.container.State = cproto.Terminated
+			j.informTaskResourcesStopped(sproto.ResourcesStopped{
+				Failure: &sproto.ResourcesRestoreError{
+					FailureType: sproto.ResourcesFailed,
+					ErrMsg:      j.firstPodExit.msg,
+					ExitCode:    (*sproto.ExitCode)(&j.firstPodExit.code),
+				},
+			})
+			return cproto.Terminated, nil
+		}
+	}
 
-	containerState, err := p.getPodState(p.pod, p.containerNames)
+	return j.container.State, nil
+}
+
+func (j *job) jobDeleted() {
+	exitCode := 1026
+	j.container.State = cproto.Terminated
+	j.informTaskResourcesStopped(sproto.ResourcesStopped{
+		Failure: &sproto.ResourcesRestoreError{
+			FailureType: sproto.ResourcesFailed,
+			ErrMsg:      "job was suddenly deleted but not by us (setting exit code 1026)",
+			ExitCode:    (*sproto.ExitCode)(&exitCode),
+		},
+	})
+}
+
+func (j *job) podStatusUpdate(updatedPod k8sV1.Pod) error {
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	// TODO(TICKET): This is how the code was before but it doesn't make sense at all.
+	// spec.NodeName is a request to schedule on a specific node, not the node each pod is on.
+	// Needs a ticket.
+	j.podNodeNames[updatedPod.Name] = updatedPod.Spec.NodeName
```

Comment:
make & link the ticket to the JIRA epic pls

Reply:
yep. i'm not planning on doing this until right before i land the pr.